src/minimization.cpp

```diff
@@ -0,0 +1,289 @@
+//===----------------------------------------------------------------------===/ /
+//
+// Part of the Utopia EDA Project, under the Apache License v2.0
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2023 ISP RAS (http://www.ispras.ru)
+//
+//===----------------------------------------------------------------------===//
+
+#include "layout.h"
+
+#include <vector>
+
+typedef std::pair<TreeNode *, TreeNode *> Edge;
+
+namespace {
+  struct AdditionalNetFeatures {
+    std::vector<std::vector<TreeNode::Id>> nodesByLayer;
+    std::vector<std::vector<TreeNode::Id>> tempNodesByLayer;
+    std::vector<std::vector<Edge>> netEdges;
+    int intersections = -1;
+    std::vector<int> accTree;
+
+    void minimizeIntersections(Net &net);
+    void setEdgesToOptimalCondition(Net &net);
+    int crossCounting(AdditionalNetFeatures &features);
+  };
+}
+
+float forwardRankDefinition(TreeNode &node, int nodeIndex, int portIndex) {
+  return nodeIndex + portIndex / (node.succ.size() + 1);
+}
+
+float backwardRankDefinition(TreeNode &node, int nodeIndex, int portIndex) {
+  int portOrderValue;
+  int maxPortIndex = node.pred.size() + node.succ.size();
+  if (portIndex <= maxPortIndex) {
+    portOrderValue = maxPortIndex - portIndex + 1;
+  } else {
+    portOrderValue = maxPortIndex + node.succ.size() - portIndex + 1;
+  }
+  return nodeIndex + portOrderValue / (node.pred.size() + 1);
+}
+
+void sortNodes(Net &net, std::vector<TreeNode::Id> &layer) {
+  std::sort(layer.begin(), layer.end(), [&net](const auto &x,
+    const auto &y) -> bool {
+    return net.getNode(x)->barycentricValue < net.getNode(y)->barycentricValue;
+  });
```

Comment:
Broken indentation, no need to pass ids by-reference
```suggestion
  std::sort(layer.begin(), layer.end(), [&net](auto x, auto y) -> bool {
    return net.getNode(x)->barycentricValue < net.getNode(y)->barycentricValue;
  });
```